src/rust/engine/logging/src/executor.rs

```diff
@@ -0,0 +1,130 @@
+use futures::Future;
+use std::sync::Arc;
+use tokio::runtime::Runtime;
+
+// TODO: It's strange that this is an exposed interface from the logging crate, rather than an
+// implementation of a trait that lives elsewhere. This can't currently be a trait because its
+// methods have generic types, which isn't allowed on traits. If we can move the API somewhere else
+// in the future, that could be nice.
+
+#[derive(Clone)]
+pub struct Executor {
+  runtime: Arc<Runtime>,
+  io_pool: futures_cpupool::CpuPool,
+}
+
+impl Executor {
+  pub fn new() -> Executor {
+    Executor {
+      runtime: Arc::new(
+        Runtime::new().unwrap_or_else(|e| panic!("Could not initialize Runtime: {:?}", e)),
+      ),
+      io_pool: futures_cpupool::CpuPool::new_num_cpus(),
+    }
+  }
+
+  ///
+  /// Drive running of a Future on a tokio Runtime as a new Task.
+  ///
+  /// The future will be driven to completion, but the result can't be accessed directly.
+  ///
+  /// This may be useful e.g. if you want to kick off a potentially long-running task, which will
+  /// notify dependees of its completion over an mpsc channel.
+  ///
+  pub fn spawn_and_ignore<F: Future<Item = (), Error = ()> + Send + 'static>(&self, future: F) {
+    self
+      .runtime
+      .executor()
+      .spawn(Self::future_with_correct_logging_context(future))
+  }
+
+  ///
+  /// Run a Future on a tokio Runtime as a new Task, and return a Future handle to it.
+  ///
+  /// The future will only be driven to completion if something drives the returned Future. If the
+  /// returned Future is dropped, the computation may be cancelled.
+  ///
+  /// This may be useful for tokio tasks which use the tokio blocking feature (unrelated to the
+  /// Executor::block_on method). When tokio blocking tasks run, they prevent progress on any
+  /// futures running in the same task. e.g. if you run f1.select(f2) and f1 and f2 are
+  /// tokio blocking futures, f1 and f2 will not run in parallel, defeating the point of select.
+  ///
+  /// On the other hand, if you run:
+  /// spawn_oneshot(f1).select(spawn_oneshot(f2))
+  /// those futures will run in parallel.
+  ///
+  /// Using spawn_oneshot allows for selecting the granularity when using tokio blocking.
+  ///
+  /// See https://docs.rs/tokio-threadpool/0.1.15/tokio_threadpool/fn.blocking.html for details of
+  /// tokio blocking.
+  ///
+  pub fn spawn_oneshot<
```

Comment:
This is unused I think?

Reply:
It is, but I'd like to keep it in as a handy utility that we'll probably start using if/when we start trying to use tokio-threadpoll's blocking again